src/communication/rpc_impl.rs

```diff
@@ -0,0 +1,486 @@
+            Some(UCode::OK) | None => {
+                // successful invocation
+                response.payload.map_or(Ok(None), |payload| {
+                    Ok(Some(UPayload::new(
+                        payload,
+                        attribs.payload_format.enum_value_or_default(),
+                    )))
+                })
+            }
+            Some(code) => {
+                // try to extract UStatus from response payload
+                let status = response.extract_protobuf().unwrap_or_else(|_e| {
+                    UStatus::fail_with_code(code, "failed to invoke service operation")
+                });
+                Err(ServiceInvocationError::from(status))
+            }
+        }
+    } else {
+        Err(ServiceInvocationError::RpcError(UStatus::fail_with_code(
+            UCode::INTERNAL,
+            "response message does not contain attributes",
+        )))
+    }
+}
+
+struct ResponseListener {
+    // request ID -> sender for response message
+    pending_requests: Mutex<HashMap<UUID, Sender<UMessage>>>,
+}
+
+impl ResponseListener {
+    fn try_add_pending_request(
+        &self,
+        reqid: UUID,
+    ) -> Result<Receiver<UMessage>, ServiceInvocationError> {
+        if let Ok(mut pending_requests) = self.pending_requests.lock() {
+            if let Entry::Vacant(entry) = pending_requests.entry(reqid) {
+                let (tx, rx) = tokio::sync::oneshot::channel();
+                entry.insert(tx);
+                Ok(rx)
+            } else {
+                Err(ServiceInvocationError::AlreadyExists(
+                    "RPC request with given ID already pending".to_string(),
+                ))
+            }
+        } else {
+            Err(ServiceInvocationError::RpcError(UStatus::fail_with_code(
+                UCode::INTERNAL,
+                "failed to add response handler",
+            )))
+        }
+    }
+
+    fn handle_response(&self, reqid: &UUID, response_message: UMessage) {
+        if let Ok(mut pending_requests) = self.pending_requests.lock() {
+            if let Some(sender) = pending_requests.remove(reqid) {
+                let _ = sender.send(response_message);
+            } else {
+                // we seem to have received a duplicate of the response message,
+                // ignoring it
+            }
+        }
+    }
+
+    fn remove_pending_request(&self, reqid: &UUID) -> Option<Sender<UMessage>> {
+        self.pending_requests
+            .lock()
+            .map_or(None, |mut pending_requests| pending_requests.remove(reqid))
+    }
+
+    #[cfg(test)]
+    fn contains(&self, reqid: &UUID) -> bool {
+        self.pending_requests
+            .lock()
+            .map_or(false, |pending_requests| {
+                pending_requests.contains_key(reqid)
+            })
+    }
+}
+
+#[async_trait]
+impl UListener for ResponseListener {
+    async fn on_receive(&self, msg: UMessage) {
+        if msg.is_response() {
```

Comment:
Sorry to harp on this, but again, I like reducing indentation / nesting :slightly_smiling_face: 

```rust
if !msg.is_response() {
  // TODO: Consider logging some sort of trace or info level message?
  return;
}

let Some(reqid) = msg
  .attributes
  .as_ref()
  .and_then(|attribs| attribs.reqid.clone().into_options()) else 
{
  // TODO: Consider logging some sort of trace or info level message?
  return;
}

self.handle_response(&reqid, msg);
```